core/src/apps/cardano/helpers/tx_builder.py

```diff
@@ -0,0 +1,209 @@
+from micropython import const
+
+from trezor.crypto import hashlib
+
+from apps.common import cbor
+
+from .cbor_hash_builder import CborHashBuilder
+from .lazy_collection import DICT, LAZY_PLACEHOLDER, LIST, LazyCollection
+from .tx_builder_state_machine import TxBuilderStateMachine
+
+TX_BODY_KEY_INPUTS = const(0)
+TX_BODY_KEY_OUTPUTS = const(1)
+TX_BODY_KEY_FEE = const(2)
+TX_BODY_KEY_TTL = const(3)
+TX_BODY_KEY_CERTIFICATES = const(4)
+TX_BODY_KEY_WITHDRAWALS = const(5)
+TX_BODY_KEY_AUXILIARY_DATA = const(7)
+TX_BODY_KEY_VALIDITY_INTERVAL_START = const(8)
+
+ASSET_GROUPS_POSITION_IN_OUTPUT = [1, 1]
+POOL_OWNERS_POSITION_IN_CERTIFICATE = [7]
+POOL_RELAYS_POSITION_IN_CERTIFICATE = [8]
+
+
+class TxBuilder:
+    def __init__(self, tx_body_map_item_count: int):
+        self.cbor_hash_builder = CborHashBuilder(
+            hashlib.blake2b(outlen=32), LazyCollection(DICT, tx_body_map_item_count)
+        )
+        self.cbor_hash_builder.update_hash()
+        self.state_machine = TxBuilderStateMachine()
+
+    def get_tx_hash(self) -> bytes:
+        return self.cbor_hash_builder.get_hash()
+
+    def start_inputs(self, inputs_count: int) -> None:
+        self.state_machine.transition(TxBuilderStateMachine.ACTION_START_INPUTS)
+        self.cbor_hash_builder.start_collection_at_key(
+            key=TX_BODY_KEY_INPUTS,
+            collection=LazyCollection(LIST, inputs_count),
+        )
+
+    def add_input(self, prev_hash: bytes, prev_index: int) -> None:
+        self.state_machine.transition(TxBuilderStateMachine.ACTION_ADD_INPUT)
+        self.cbor_hash_builder.add_item((prev_hash, prev_index))
+
+    def finish_inputs(self) -> None:
+        self.state_machine.transition(TxBuilderStateMachine.ACTION_FINISH_INPUTS)
+        self.cbor_hash_builder.finish_collection()
+
+    def start_outputs(self, outputs_count: int) -> None:
+        self.state_machine.transition(TxBuilderStateMachine.ACTION_START_OUTPUTS)
+        self.cbor_hash_builder.start_collection_at_key(
+            key=TX_BODY_KEY_OUTPUTS,
+            collection=LazyCollection(LIST, outputs_count),
+        )
+
+    def add_simple_output(self, amount: int, address: bytes) -> None:
+        self.state_machine.transition(TxBuilderStateMachine.ACTION_ADD_SIMPLE_OUTPUT)
+        self.cbor_hash_builder.add_item((address, amount))
+
+    def add_output_with_tokens(self, amount: int, address: bytes) -> None:
+        self.state_machine.transition(
+            TxBuilderStateMachine.ACTION_ADD_OUTPUT_WITH_TOKENS
+        )
+        self.cbor_hash_builder.prepare_nested_collections(
+            (address, [amount, LAZY_PLACEHOLDER])
```

Comment:
We need a list here (and in `add_pool_registration_certificate`) to be able to replace the placeholder.
Should we try to hide this detail somehow? Like doing some automatic tuple -> list conversion in `_set_nested_element`?
https://github.com/vacuumlabs/trezor-firmware/blob/f50d9a913d49ac876677db6626eaa2545354f342/core/src/apps/cardano/helpers/cbor_hash_builder.py#L88-L94

Reply:
I guess the list is ok. At least I wasn't able to come up with a suitable alternative. But I'd add a comment here to explain it.

I'm not sure the tuple -> list conversion could work since I think we need the same references to the structures, no?